docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,561 @@
+
+The authors propose to use a custom representation for states, sub-states and transitions.
+
+Consider the following representation:
+
+```yaml
+transitions:
+- state: <name of the state that the etcd-member has transitioned to>
+  subState: <name of the sub-state if any>
+  reason: <reason code for the transition>
+  transitionTime: <time of transition to this state>
+  message: <detailed message if any>
+```
+
+As an example, consider the following transitions which represent addition of an etcd-member during scale-up of an etcd cluster, followed by a restart of the etcd-member which detects a corrupt DB:
+
+```yaml
+status:
+  transitions:
+  - state: New
+    subState: New
+    reason: ClusterScaledUp
+    transitionTime: "2023-07-17T05:00:00Z"
+    message: "New member added due to etcd cluster scale-up"
+  - state: Starting
+    subState: PendingLearner
+    reason: WaitingToJoinAsLearner
+    transitionTime: "2023-07-17T05:00:30Z"
+    message: "Waiting to join the cluster as a learner"
+  - state: Starting
+    subState: Learner
+    reason: JoinedAsLearner
+    transitionTime: "2023-07-17T05:01:20Z"
+    message: "Joined the cluster as a learner"
+  - state: Started
+    subState: Follower
+    reason: PromotedAsVotingMember
+    transitionTime: "2023-07-17T05:02:00Z"
+    message: "Now in sync with leader, promoted as voting member"
+  - state: Initializing
+    subState: DBValidationFull
+    reason: DetectedPreviousUncleanExit
+    transitionTime: "2023-07-17T08:00:00Z"
+    message: "Detected previous unclean exit, requires full DB validation"
+  - state: New
+    subState: New
+    reason: DBCorruptionDetected
+    transitionTime: "2023-07-17T08:01:30Z"
+    message: "Detected DB corruption during initialization, removing member from cluster"
+  - state: Starting
+    subState: PendingLearner
+    reason: WaitingToJoinAsLearner
+    transitionTime: "2023-07-17T08:02:10Z"
+    message: "Waiting to join the cluster as a learner"
+  - state: Starting
+    subState: Learner
+    reason: JoinedAsLearner
+    transitionTime: "2023-07-17T08:02:20Z"
+    message: "Joined the cluster as a learner"
+  - state: Started
+    subState: Follower
+    reason: PromotedAsVotingMember
+    transitionTime: "2023-07-17T08:04:00Z"
+    message: "Now in sync with leader, promoted as voting member"
+```
+
+##### Reason Codes
+
+The authors propose the following list of possible reason codes for transitions. This list is not exhaustive, and can be further enhanced to capture any new transitions in the future.
+
+| Reason                                | Transition From State (SubState)                             | Transition To State (SubState)         |
+| ------------------------------------- | ------------------------------------------------------------ | -------------------------------------- |
+| `ClusterCreated`                      | nil                                                          | New                                    |
```

Comment:
`ClusterCreated` as the reason is a little misleading as it depicts a completed activity more in line with final state. 
I was wondering if it will be better to change it to something in present or present-continuous tense like `CreateCluster` or `CreatingCluster`.

Reply:
Yes this was also the comment during an earlier discussion as well. I have now proposed a change in the name. PTAL.

Reply:
Nice I missed that clarity of `ClusterScaleup` and `NewSingleNodeClusterCreated`. This looks more apt.